File: topological-sort/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        store = {i:[] for i in range(numCourses)}
        for crs, preq in prerequisites:
            store[crs].append(preq)
        
        visited = set()
        def dfs(crs):

            if crs in visited:
                return False
            if store[crs] == []:
                return True
            visited.add(crs)
            for nei in store[crs]:
                if not dfs(nei): return False
            store[crs] = []
            visited.remove(crs)
            return True

        for crs in range(numCourses):
            if not dfs(crs): return False
        return True 
        # visited = set()
        # completed = set()
        
        # def dfs(crs):
        #     if crs in visited:
        #         return False
        #     if crs in completed:
        #         return True
            
        #     visited.add(crs)
        #     for nei in store[crs]:
        #         if not dfs(nei):
        #             return False
        #     visited.remove(crs)
        #     completed.add(crs)
        #     return True

        # for crs in range(numCourses):
        #     if not dfs(crs):
        #         return False
        # return True

File: breadth-first-search/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // Build the graph
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] prerequisite : prerequisites) {
            graph.get(prerequisite[1]).add(prerequisite[0]);
        }

        // Array to track the state of each node (0 = unvisited, 1 = visiting, 2 = visited)
        int[] visited = new int[numCourses];

        // Perform DFS for each course
        for (int i = 0; i < numCourses; i++) {
            if (visited[i] == 0) {
                if (hasCycle(graph, visited, i)) {
                    return false; // If a cycle is detected, return false
                }
            }
        }

        return true; // If no cycles are detected, return true
    }

    // Helper function to perform DFS and detect cycles
    private boolean hasCycle(List<List<Integer>> graph, int[] visited, int course) {
        if (visited[course] == 1) {
            return true; // A cycle is detected
        }
        if (visited[course] == 2) {
            return false; // The node has already been visited and no cycle was found
        }

        visited[course] = 1; // Mark the node as visiting

        // Visit all the neighbors
        for (int neighbor : graph.get(course)) {
            if (hasCycle(graph, visited, neighbor)) {
                return true; // If a cycle is detected in any neighbor, return true
            }
        }

        visited[course] = 2; // Mark the node as visited
        return false; // No cycle detected
    }
}